smolyay/surrogate.py

```diff
@@ -1,546 +1,729 @@
+        if self.regularization != value:
+            self._regularization = value
+            self._needs_fit = False
+
+    @property
+    def basis_sets(self):
+        """list of BasisFunctionSet: the set of basis functions for the terms."""
+        return self._basis_sets
+
+    def predict(self, X):
         """Evaluate surrogate at a given input.
 
-        Parameter
-        ---------
-        x: list (for dimension > 1), or float (for dimension = 1)
-            Input.
+        Parameters
+        ----------
+        X: array-like with shape (n_samples, num_dimensions) or list of object
+            Points at which the model is evaluated.
 
         Returns
         -------
-        float
+        ndarray of shape (n_samples,) or (n_samples, num_dimensions)
             Surrogate output at x.
 
         Raises
         ------
-        ValueError
+        RuntimeError
             For surrogate to be evaluated, function needs to be trained.
         IndexError
-            Input of the surrogate must be of length of the
-            function's dimensionality.
+            Input must be 2D array with shape (n_samples, num_dimensions).
         ValueError
             Input must lie in domain of surrogate.
+
         """
-        if self._coefficients is None:
-            raise ValueError('Function needs training!')
-        input_shape = numpy.shape(x)
-        x = numpy.array(x, copy=False, ndmin=2)
-        if self.dimension == 1:
-            if len(input_shape) <= 1:
-                # the cast to 2d puts these in wrong order, so transpose
-                x = x.T
-            elif x.shape[-1] > 1:
-                x = x[..., numpy.newaxis]
-        
-        if x.shape[-1] != self.dimension:
-            raise IndexError("Input must match dimension of domain")
-        if self.dimension > 1:
-            oob = any(numpy.any(x[..., i] < self.domain[i][0]) or 
-                    numpy.any(x[..., i] > self.domain[i][1]) 
-                    for i in range(self.dimension))
+        # validate inputs
+        if not self._needs_fit:
+            raise RuntimeError("Model must be fit!")
+        X = numpy.array(X, ndmin=2, copy=None)
+        self._assert_in_domain(X)
+
+        # create lookup table and solve for all the basis functions
+        lookup_table = [
+            self.basis_sets[dim](X[:, dim], self.domain[dim])
+            for dim in range(self.num_dimensions)
+        ]
+
+        # use lookup table to combine terms
+        answer = numpy.ones(len(X)) * self._integration_constant
+        for ic, coeff in zip(self._index_combinations, self._coefficients):
+            answer += numpy.real(
+                coeff
+                * numpy.prod(
+                    [lookup_table[dim][ic[dim], :] for dim in range(len(ic))], axis=0
+                )
+            )
+
+        # return results
+        if len(X) == 1:
+            return answer[0]
```

Comment:
I wanted to check, is this the behavior of scikit learn? This will return a scalar if `X` has only one row, but it could also return a one-element array.

In the old code, I think we did some deduction: for a multidimensional input space, if a 1d array was supplied, then a scalar was returned. If a 2d array was supplied, then an array was returned.

Reply:
It does return a 1D element array with shape (n_samples,). I'll make the adjustments.